text/1021-rewrites.rst

```diff
@@ -0,0 +1,210 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2023-01-12
+    Authors: Michel J. Sullivan <sully@msully.net>
+
+===========================
+RFC 1021: Mutation rewrites
+===========================
+
+This RFC proposes a mechanism for specifying rewrite rules that are
+applied to properties and links on INSERT and UPDATE of objects.
+
+They can be thought of as a generalization of ``default``.
+
+Motivation
+==========
+
+It is sometimes useful to have a pointer in an object automatically updated
+whenever the object has an ``UPDATE`` performed on it.
+
+The most straightforward example of this is having a
+``modification_time`` field that tracks when an object was last
+modified.
+
+It is also valuable when (probably for performance reasons) you wish
+to maintain a field that is derived from other object
+state. (For example, maintaining a count of friends.)
+
+Note
+====
+
+This is a companion to
+`RFC 1020 <https://github.com/edgedb/rfcs/blob/master/text/1020-triggers.rst>`_,
+in that they work together to cover most of the desired use cases that motivated
+a previous `previous version triggers RFC <https://github.com/edgedb/rfcs/pull/70>`_
+
+Specification
+=============
+
+Synopsis::
+
+    type <type-name> "{"
+        { property | link } <name> -> <type> "{"
+            rewrite { insert | update } [, ... ] using (<expr>)
+        "}"
+    "}"
+
+We allow pointer definitions to be augmented with a ``rewrite`` rule.
+
+Expression interpretation
+-------------------------
+In ``<expr>``, ``__subject__`` (and the implicit path prefix) will be
+bound to the new value object being inserted or updated.
+
+In purely ``update`` rewrite rules, ``__old__`` will refer to the old
+version of the object.
+
+The special variable ``__fields__`` will be bound to all of the
+properties of ``<type-name>`` that were explicitly provided in the
+``insert`` or ``update``. This allows rewrite rules to distinguish
+between an explicitly provided empty set and no value being specified.
+
+(Whether ``<type-name>`` should refer to the object being mutated is
+kind of fraught; we leave that undecided for now. Some discussion of
+that
+`here
+<https://github.com/edgedb/edgedb/issues/4142#issuecomment-1386287450>_`.)
+
+Restrictions
+------------
+Each pointer can have one ``insert`` rewrite rule and one ``update``
+rewrite rule. They can be specified together if desired.
+
+Specifying a rewrite rule on a pointer is not allowed if an ancestor
+has already specified that kind of rewrite rule for the pointer.
+(Alternately: multiple rewrite rules are applied in MRO order?)
```

Comment:
what about the first rewrite rule found in MRO order is applied? This seems more intuitive to me.

If I think about this in terms of methods and object inheritance, which is probably the typical user's background then the first option sounds like "you can't define a method if an ancestor has already defined it" which doesn't seem helpful. The second option sounds like "every definition of a given method will be called in MRO order" which sounds even worse.

There are probably a lot of things I'm missing here though...

Reply:
I agree that overloading makes more sense here and is consistent with how you can overload a `default`.

Reply:
I'm definitely +1 on overloading as the mechanism here. We opt for (you cannot overload this) only when we can't figure out how to do that reasonably and often it's a temporary restriction (because it's forward-compatible with overloading).